src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineView.xaml.cs

```diff
@@ -0,0 +1,183 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Windows;
+using System.Windows.Controls;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServices.Implementation;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineView.xaml
+    /// </summary>
+    internal partial class DocumentOutlineView : UserControl, IVsCodeWindowEvents, IDisposable
+    {
+        private readonly DocumentOutlineViewModel _viewModel;
+        private readonly VisualStudioCodeWindowInfoService _visualStudioCodeWindowInfoService;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+        private readonly ComEventSink _codeWindowEventsSink;
+
+        public DocumentOutlineView(
+            DocumentOutlineViewModel viewModel,
+            VisualStudioCodeWindowInfoService visualStudioCodeWindowInfoService,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            _viewModel = viewModel;
+            _visualStudioCodeWindowInfoService = visualStudioCodeWindowInfoService;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            DataContext = _viewModel;
+            InitializeComponent();
+
+            // We don't think the shell is initialized lazily, so we'll Debug.Fail(), but if it was we'd still
+            // see the view created later so this will still function.
+            if (ErrorHandler.Failed(codeWindow.GetPrimaryView(out var primaryTextView)))
+                Debug.Fail("GetPrimaryView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Failed(StartTrackingView(primaryTextView)))
+                Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Succeeded(codeWindow.GetSecondaryView(out var secondaryTextView)))
+            {
+                if (ErrorHandler.Failed(StartTrackingView(secondaryTextView)))
+                    Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+            }
+
+            _codeWindowEventsSink = ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
+        }
+
+        private int StartTrackingView(IVsTextView textView)
+        {
+            var wpfTextView = _editorAdaptersFactoryService.GetWpfTextView(textView);
```

Comment:
i'd add threading asserts to all these methods.

Reply:
(applies to all methods).

Reply:
We have not internal or public members in this type. If we are ever called from a background thread it would be a bug in WPF itself.

Reply:
our basic pattern in roslyn ide is if there are threading constraints we enforce them with real (runtime/release) asserts.  That way when things get screwed up (since we can call these things ourselves) it is caught immediately :)   This has saved us a *lot* of pain with maintaining UI affinitized code.